sql/parser/eval.go

```diff
@@ -1214,11 +1434,61 @@ func evalCastExpr(expr *CastExpr) (Datum, error) {
 		}
 		return s, nil
 
-		// TODO(pmattis): unimplemented.
-		// case *DecimalType:
-		// case *DateType:
-		// case *TimestampType:
+	// TODO(pmattis): unimplemented.
+	// case *DecimalType:
+	case *DateType:
+		switch d.(type) {
+		case DString:
+			return ParseDate(d.(DString))
+		}
+	case *TimestampType:
+		switch d.(type) {
+		case DString:
+			return ParseTimestamp(d.(DString))
+		}
+	case *IntervalType:
+		switch d.(type) {
+		case DString:
+			d, err := time.ParseDuration(string(d.(DString)))
+			return DInterval{Duration: d}, err
+		}
 	}
 
 	return DNull, fmt.Errorf("invalid cast: %s -> %s", d.Type(), expr.Type)
 }
+
+const (
+	dateFormat                    = "2006-01-02"
+	timestampFormat               = "2006-01-02 15:04:05.999999999"
```

Comment:
Don't you also need to support timestamp formats without nanoseconds? Or does that "just work" with these format strings?